vpr/src/place/cut_spreader.h

```diff
@@ -0,0 +1,259 @@
+     * 			Strict_legalize must be run after for legal placement result, and for legal placement to
+     * 			be passed to annealer through vpr_ctx.
+     *
+     * 			Input placement is passed by data members (blk_locs) in analytic_placer
+     *
+     * @return	result placement is passed to strict legalizer by modifying blk_locs in analytic_placer
+     */
+    void cutSpread();
+
+    /*
+     * @brief: 	Greedy strict legalize using algorithm described in algorithm overview above.
+     *
+     * 			Input illegal placement from data members (blk_locs) in analytic_placer
+     *
+     * @return	result placement is passed to annealer by modifying vpr_ctx.mutable_placement()
+     */
+    void strict_legalize();
+
+  private:
+    // marker for blocks that wasn't solved (blocks of different type, macro members)
+    int32_t dont_solve = std::numeric_limits<int32_t>::max();
+
+    // pointer to analytic_placer to access its data members
+    AnalyticPlacer* ap;
+
+    // block type to legalize
+    t_logical_block_type_ptr blk_type;
+
+    // struct that specifies the region occupied by a macro
+    struct MacroExtent {
+        int x0, y0, x1, y1;
+    };
+
+    // struct describing regions on FPGA to cut_spread
+    struct SpreaderRegion {
+        int id;              // index of regions in regions vector
+        int x0, y0, x1, y1;  // extent of the region
+        int n_blks, n_tiles; // number of blocks and compatible tiles
+        bool overused(float beta) const {
+            // determines whether region is overutilized:
+            // overused = (Occupancy / Capacity) > beta
+            if (n_blks > beta * n_tiles) return true;
+            return false;
+        }
+    };
+
+    // Utilization of each tile, indexed by x, y
+    std::vector<std::vector<int>> occupancy;
+
+    // Region ID of each tile, indexed by x, y. -1 if not covered by any region
+    std::vector<std::vector<int>> groups;
+
+    // Extent of macro at x, y location. If blk is not in any macros, MacroExtent only covers a single locations
+    std::vector<std::vector<MacroExtent>> macros;
+
+    // List of logic blocks of blk_type at x, y location, indexed by x, y
+    std::vector<std::vector<std::vector<ClusterBlockId>>> blks_at_location;
+
+    // List of sub_tiles compatible with blk_type at x, y location, indexed by x, y
+    std::vector<std::vector<std::vector<t_pl_loc>>> ft;
+
+    // List of all SpreaderRegion, index of vector members is the id of the region
+    std::vector<SpreaderRegion> regions;
+
+    // List of all merged_regions, these regions are merged in larger regions and should be skipped when
+    // recursively cur_spreading. Each entry is the region's ID, which is also the index into the regions vector
+    std::unordered_set<int> merged_regions;
+
+    // Lookup of macro's extent by block ID. If block is a single block, MacroExtent contains only 1 tile location
+    std::map<ClusterBlockId, MacroExtent> blk_extents;
```

Comment:
vtr::vector_map<ClusterBlockId, MacroExtent> or 
vtr::vector_map<ClusterBlockID, vtr::Rect>  // even better
would be faster and smaller.